File: App/UI/Main/Tabs/ButtonTab/RemapTable/KeyCaptureView/KeyCaptureView.m
//
// --------------------------------------------------------------------------
// MFKeystrokeCaptureView.m
// Created for Mac Mouse Fix (https://github.com/noah-nuebling/mac-mouse-fix)
// Created by Noah Nuebling in 2021
// Licensed under the MMF License (https://github.com/noah-nuebling/mac-mouse-fix/blob/master/License)
// --------------------------------------------------------------------------
//

#import "KeyCaptureView.h"
#import "AppDelegate.h"
#import "UIStrings.h"
#import <Carbon/Carbon.h>
#import "MFMessagePort.h"
#import "NSView+Additions.h"
#import "Mac_Mouse_Fix-Swift.h"

@interface KeyCaptureView ()

- (IBAction)backgroundButton:(id)sender;

@end

@implementation KeyCaptureView {
    
    BOOL _isCapturing;
    
    CaptureHandler _captureHandler;
    CancelHandler _cancelHandler;
    
    id _localEventMonitor;
    
    NSDictionary *_attributesFromIB;
}

#pragma mark - (Pseudo) Properties

- (void)setCoolString:(NSString *)string {
    
    NSAttributedString *attributedString = [[NSAttributedString alloc] initWithString:string attributes:_attributesFromIB];
    
    self.textStorage.attributedString = attributedString;
}

#pragma mark - Setup

- (void)setupWithCaptureHandler:(CaptureHandler)captureHandler
                   cancelHandler:(CancelHandler)cancelHandler {
    
    DDLogDebug(@"Setting up keystroke capture view");
    
    self.delegate = self;
    _captureHandler = captureHandler;
    _cancelHandler = cancelHandler;
    
    dispatch_async(dispatch_get_main_queue(), ^{
        [MainAppState.shared.window makeFirstResponder:self];
    });
    // ^ This view is being drawn by the tableView. Using dispatch_async makes it execute after the tableView is done drawing preventing a crash
    
}

#pragma mark - Lifecycle and drawing

- (void)awakeFromNib {
    
    if (_attributesFromIB == nil) {
        _attributesFromIB = [self.attributedString attributesAtIndex:0 effectiveRange:nil];
    }
}

- (void)drawEmptyAppearance { // Not really drawing in the NSFillRect sense, probably a bad name
    
    self.coolString = NSLocalizedString(@"type-shortcut-prompt", @"");
    self.textColor = NSColor.placeholderTextColor;
    
    [self selectAll:nil];
}

- (void)dealloc {
    
    [NSNotificationCenter.defaultCenter removeObserver:self];
}

#pragma mark keyCaptureModeFeedback

+ (void)handleKeyCaptureModeFeedbackWithPayload:(NSDictionary *)payload isSystemDefinedEvent:(BOOL)isSystem {
    
    /// Find keyCaptureField instance in remapsTable
    
    NSTableView *remapsTable = MainAppState.shared.remapTable;
    NSInteger effectColumn = [remapsTable columnWithIdentifier:@"effect"];
    NSMutableIndexSet *indexes = [NSMutableIndexSet new];
    for (int r = 0; r < remapsTable.numberOfRows; r++) {
        NSView *effectView = [MainAppState.shared.remapTable viewAtColumn:effectColumn row:r makeIfNecessary:NO];
        if ([effectView.identifier isEqual:@"keyCaptureCell"]) {
            [indexes addIndex:r];
        }
    }
    assert(indexes.count <= 1);
    if (indexes.count == 0) return;
    
    NSTableCellView *keyCaptureCell = [MainAppState.shared.remapTable viewAtColumn:effectColumn row:indexes.firstIndex makeIfNecessary:NO];
    KeyCaptureView *keyCaptureView = (KeyCaptureView *)[keyCaptureCell nestedSubviewsWithIdentifier:@"keyCaptureView"].firstObject;
    
    /// Send payload to found instance
    
    [keyCaptureView handleKeyCaptureModeFeedbackWithPayload:payload isSystemDefinedEvent:isSystem];
}

- (void)handleKeyCaptureModeFeedbackWithPayload:(NSDictionary *)payload isSystemDefinedEvent:(BOOL)isSystem {
    
    _isCapturing = NO; /// Helper disabled keyCaptureMode after sending payload
    
    CGKeyCode keyCode = USHRT_MAX;
    MFSystemDefinedEventType type = UINT_MAX;
    CGEventFlags flags;
    
    if (isSystem) {
        type = ((NSNumber *)payload[@"systemEventType"]).unsignedIntValue;
        flags = ((NSNumber *)payload[@"flags"]).unsignedLongValue;
    } else {
        keyCode = ((NSNumber *)payload[@"keyCode"]).unsignedShortValue;
        flags = ((NSNumber *)payload[@"flags"]).unsignedLongValue;
    }
    
    [MainAppState.shared.window makeFirstResponder:nil]; /// Important to call this before capture handler, otherwise `resignFirstResponder:` (our teardown function) isn't called
    
    _captureHandler(keyCode, type, flags); /// This should undraw the view
}

#pragma mark FirstResponderStatus handlers

- (BOOL)becomeFirstResponder {
    
    DDLogDebug(@"BECOME FIRST RESPONDER");
    
    BOOL superAccepts = [super becomeFirstResponder];
    
    if (superAccepts) {
        
        _isCapturing = YES;
        
        // If the window goes to the background, resign key
        [NSNotificationCenter.defaultCenter addObserverForName:NSWindowDidResignKeyNotification object:MainAppState.shared.window queue:nil usingBlock:^(NSNotification * _Nonnull note) {
                    [MainAppState.shared.window makeFirstResponder:nil];
        }];
        
        [MFMessagePort sendMessage:@"enableKeyCaptureMode" withPayload:@"" waitForReply:NO];
        /// ^ Do actual capturing in helper app because it already has permissions to stop captured events from being sent to other apps
        
        [self drawEmptyAppearance];
        
        _localEventMonitor = [NSEvent addLocalMonitorForEventsMatchingMask:(NSEventMaskFlagsChanged | NSEventMaskKeyDown | NSEventMaskLeftMouseDown) handler:^NSEvent * _Nullable(NSEvent * _Nonnull event) {
            CGEventRef e = event.CGEvent;
            /// User is playing around with modifier keys
            
            if (event.type == NSEventTypeFlagsChanged) {
                CGEventFlags flags = CGEventGetFlags(e);
                
                NSString *modString = [UIStrings getKeyboardModifierString:flags];
                if (modString.length > 0) {
                    self.coolString = modString;
                } else {
                    [self drawEmptyAppearance];
                }
            } else if (event.type == NSEventTypeKeyDown) {
//                assert(!self->_isCapturing); // _isCapturing should be set to NO by `handleKeyCaptureModeFeedbackWithPayload:` before this is executed.
            } else if (event.type == NSEventTypeLeftMouseDown) {
                /// If the user clicks anything, resign key. -> To prevent weird states. E.g. where Mac Mouse Fix is disabled while the field is still up
                [MainAppState.shared.window makeFirstResponder:nil];
            }
            
            /// Set the `RECORDING_MODE` flag when you want to record video demos, so that cleanshot can properly highlight when a button or keyboard key is pressed. Edit: This doesn't seem to work here.
#if RECORDING_MODE
            return event;
#endif
            return nil;
        }];
    }
    
    return superAccepts;
}
- (BOOL)resignFirstResponder {

    DDLogDebug(@"RESIGN FIRST RESPONDER");
    
    BOOL superResigns = [super resignFirstResponder];

    if (superResigns) {
        
        [MFMessagePort sendMessage:@"disableKeyCaptureMode" withPayload:nil waitForReply:NO];
        [NSEvent removeMonitor:_localEventMonitor];
        _localEventMonitor = nil; /// Otherwise crashes on macOS 10.13 and 10.14. Didn't test other versions.
        _cancelHandler();
    }
    return superResigns;
}

#pragma mark - Disable MouseDown and mouseover cursor

- (void)mouseDown:(NSEvent *)event {
    /// Ignore
}
- (void)mouseMoved:(NSEvent *)event {
    [NSCursor.arrowCursor set]; /// Prevent text insertion cursor from appearing on mouseover
}
- (void)scrollWheel:(NSEvent *)event {
    [NSCursor.arrowCursor set]; /// Prevent text insertion cursor from appearing on scroll
}

- (IBAction)backgroundButton:(id)sender {
}
@end

File: App/UI/Main/Tabs/ButtonTab/RemapTable/KeyCaptureView/KeyCaptureView.swift
//
// --------------------------------------------------------------------------
// KeyCaptureView.swift
// Created for Mac Mouse Fix (https://github.com/noah-nuebling/mac-mouse-fix)
// Licensed under the MMF License (https://github.com/noah-nuebling/mac-mouse-fix/blob/master/License)
// --------------------------------------------------------------------------
//

import Cocoa
import Carbon

public typealias CaptureHandler = (CGKeyCode, UInt32, CGEventFlags) -> Void
public typealias CancelHandler = () -> Void

@objc(KeyCaptureView)
public class KeyCaptureView: NSTextView, NSTextViewDelegate {
    
    private var _isCapturing: Bool = false
    private var _captureHandler: CaptureHandler?
    private var _cancelHandler: CancelHandler?
    private var _localEventMonitor: Any?
    private var _attributesFromIB: [NSAttributedString.Key: Any]?
    private var _windowResignObserver: Any?
    
    public override var acceptsFirstResponder: Bool {
        return true
    }
    
    @objc public func setCoolString(_ string: String) {
        let attrs = _attributesFromIB ?? [:]
        let attributedString = NSAttributedString(string: string, attributes: attrs)
        self.textStorage?.setAttributedString(attributedString)
    }
    
    @objc public func setup(withCaptureHandler captureHandler: @escaping CaptureHandler,
                             cancelHandler: @escaping CancelHandler) {
        DDLogDebug("Setting up keystroke capture view")
        self.delegate = self
        self._captureHandler = captureHandler
        self._cancelHandler = cancelHandler
        
        DispatchQueue.main.async {
            MainAppState.shared.window?.makeFirstResponder(self)
        }
    }
    
    public override func awakeFromNib() {
        super.awakeFromNib()
        if _attributesFromIB == nil {
            let attrString = self.attributedString()
            if attrString.length > 0 {
                _attributesFromIB = attrString.attributes(at: 0, effectiveRange: nil)
            }
        }
    }
    
    public func drawEmptyAppearance() {
        self.setCoolString(MFLocalizedString("type-shortcut-prompt", comment: ""))
        self.textColor = .placeholderTextColor
        self.selectAll(nil)
    }
    
    @objc public static func handleKeyCaptureModeFeedback(withPayload payload: [AnyHashable: Any], isSystemDefinedEvent isSystem: Bool) {
        guard let remapsTable = MainAppState.shared.remapTable else { return }
        let effectColumn = remapsTable.column(withIdentifier: NSUserInterfaceItemIdentifier("effect"))
        if effectColumn == -1 { return }
        let indexes = NSMutableIndexSet()
        for r in 0..<remapsTable.numberOfRows {
            if let effectView = remapsTable.view(atColumn: effectColumn, row: r, makeIfNecessary: false),
               effectView.identifier == NSUserInterfaceItemIdentifier("keyCaptureCell") {
                indexes.add(r)
            }
        }
        assert(indexes.count <= 1)
        if indexes.count == 0 { return }
        
        if let keyCaptureCell = remapsTable.view(atColumn: effectColumn, row: indexes.firstIndex, makeIfNecessary: false) {
            let subviews = keyCaptureCell.nestedSubviews(withIdentifier: NSUserInterfaceItemIdentifier("keyCaptureView"))
            if let keyCaptureView = subviews.first as? KeyCaptureView {
                keyCaptureView.handleKeyCaptureModeFeedback(withPayload: payload, isSystemDefinedEvent: isSystem)
            }
        }
    }
    
    @objc public func handleKeyCaptureModeFeedback(withPayload payload: [AnyHashable: Any], isSystemDefinedEvent isSystem: Bool) {
        _isCapturing = false
        
        var keyCode: CGKeyCode = CGKeyCode(UInt16.max)
        var type: UInt32 = UInt32(UInt32.max)
        var flags: CGEventFlags = []
        
        if isSystem {
            if let typeNum = payload["systemEventType"] as? NSNumber {
                type = typeNum.uint32Value
            }
            if let flagsNum = payload["flags"] as? NSNumber {
                flags = CGEventFlags(rawValue: flagsNum.uint64Value)
            }
        } else {
            if let codeNum = payload["keyCode"] as? NSNumber {
                keyCode = codeNum.uint16Value
            }
            if let flagsNum = payload["flags"] as? NSNumber {
                flags = CGEventFlags(rawValue: flagsNum.uint64Value)
            }
        }
        
        MainAppState.shared.window?.makeFirstResponder(nil)
        _captureHandler?(keyCode, type, flags)
    }
    
    public override func becomeFirstResponder() -> Bool {
        DDLogDebug("BECOME FIRST RESPONDER")
        let superAccepts = super.becomeFirstResponder()
        DDLogDebug("superAccepts: \(superAccepts)")
        
        _isCapturing = true
        
        if _windowResignObserver == nil {
            _windowResignObserver = NotificationCenter.default.addObserver(
                forName: NSWindow.didResignKeyNotification,
                object: MainAppState.shared.window,
                queue: nil) { [weak self] _ in
                    MainAppState.shared.window?.makeFirstResponder(nil)
                }
        }
        
        _ = MFMessagePort.sendMessage("enableKeyCaptureMode", withPayload: "" as NSObject, waitForReply: false)
        self.drawEmptyAppearance()
        
        if _localEventMonitor == nil {
            _localEventMonitor = NSEvent.addLocalMonitorForEvents(matching: [.flagsChanged, .keyDown, .leftMouseDown]) { [weak self] event in
                guard let self = self else { return event }
                guard let e = event.cgEvent else { return event }
                
                if event.type == .flagsChanged {
                    let flags = e.flags
                    let modString = UIStrings.getKeyboardModifierString(flags)
                    if modString.count > 0 {
                        self.setCoolString(modString)
                    } else {
                        self.drawEmptyAppearance()
                    }
                } else if event.type == .keyDown {
                    // Ignore
                } else if event.type == .leftMouseDown {
                    MainAppState.shared.window?.makeFirstResponder(nil)
                }
                
                return nil
            }
        }
        
        return true
    }
    
    public override func resignFirstResponder() -> Bool {
        DDLogDebug("RESIGN FIRST RESPONDER")
        let superResigns = super.resignFirstResponder()
        DDLogDebug("superResigns: \(superResigns)")
        
        _ = MFMessagePort.sendMessage("disableKeyCaptureMode", withPayload: nil, waitForReply: false)
        if let monitor = _localEventMonitor {
            NSEvent.removeMonitor(monitor)
            _localEventMonitor = nil
        }
        if let obs = _windowResignObserver {
            NotificationCenter.default.removeObserver(obs)
            _windowResignObserver = nil
        }
        _cancelHandler?()
        return true
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
        if let obs = _windowResignObserver {
            NotificationCenter.default.removeObserver(obs)
        }
        if let monitor = _localEventMonitor {
            NSEvent.removeMonitor(monitor)
        }
    }
    
    public override func mouseDown(with event: NSEvent) {
        // Ignore
    }
    
    public override func mouseMoved(with event: NSEvent) {
        NSCursor.arrow.set()
    }
    
    public override func scrollWheel(with event: NSEvent) {
        NSCursor.arrow.set()
    }
    
    @IBAction func backgroundButton(_ sender: Any) {
        // Ignore
    }
}
